File: Portfolio_Frontend/src/utils/latexSync.ts
import { PaperItem } from '../context/PortfolioContext';

export const escapeLatex = (text: string): string => {
    if (!text) return '';
    return text
        .replace(/(?<!\\)&/g, '\\&')
        .replace(/(?<!\\)%/g, '\\%')
        .replace(/(?<!\\)#/g, '\\#')
        .replace(/(?<!\\)\$/g, '\\$');
};

export const formatAuthorsLatex = (authorsStr: string): string => {
    if (!authorsStr) return '\\textbf{Shah Abdul Mazid}';
    let clean = escapeLatex(authorsStr);
    const patterns = [
        /(?<!\\textbf\{)Shah Abdul Mazid(?!\})/g,
        /(?<!\\textbf\{)Mazid, S\.\s*A\.(?!\})/g,
        /(?<!\\textbf\{)Mazid, S\.A\.(?!\})/g,
    ];
    for (const pat of patterns) {
        clean = clean.replace(pat, '\\textbf{Shah Abdul Mazid}');
    }
    return clean;
};

export const formatPaperLatex = (paper: PaperItem): string => {
    const title = escapeLatex(paper.title?.trim() || '');
    const authors = formatAuthorsLatex(paper.authors?.trim() || '');
    const venue = escapeLatex(paper.venue?.trim() || '');
    const publisher = escapeLatex(paper.publisher?.trim() || '');
    const year = String(paper.year || '').trim();
    const doi = String(paper.doi || '').trim();
    const link = String(paper.link || '').trim();

    const lines: string[] = [`\\textbf{${title}}\\\\`];

    if (authors) {
        lines.push(`\\textbf{Authors:} ${authors}\\\\[-0.5mm]`);
    }

    const venueParts: string[] = [];
    if (venue) venueParts.push(`\\textit{${venue}}`);
    if (publisher) venueParts.push(publisher);
    if (year && !venue.includes(year) && !publisher.includes(year)) {
        venueParts.push(year);
    }

    if (venueParts.length > 0) {
        lines.push(`${venueParts.join(', ')}.\\\\[-1mm]`);
    }

    if (doi) {
        const cleanDoi = doi.replace('https://doi.org/', '').trim();
        const escapedDoi = cleanDoi.replace(/_/g, '\\_');
        lines.push(`DOI: \\href{https://doi.org/${cleanDoi}}\n{${escapedDoi}}`);
    } else if (link) {
        lines.push(`Link: \\href{${link}}{View Publication}`);
    }

    return lines.join('\n');
};

export const generatePublicationsLatex = (papers: PaperItem[] = []): string => {
    if (!papers || papers.length === 0) {
        return '\\section{Publication}\n\n';
    }

    // Prioritize papers where author is Shah Abdul Mazid / Mazid, or include all papers
    const mazidPapers = papers.filter(
        p => !p.authors || p.authors.toLowerCase().includes('mazid') || p.authors.toLowerCase().includes('shah')
    );
    const selectedPapers = mazidPapers.length > 0 ? mazidPapers : papers;

    const sorted = [...selectedPapers].sort((a, b) => {
        const yA = parseInt(String(a.year || '0'), 10) || 0;
        const yB = parseInt(String(b.year || '0'), 10) || 0;
        return yB - yA;
    });

    const entries = sorted.map(formatPaperLatex);
    return `\\section{Publication}\n\n${entries.join('\n\n\\vspace{1mm}\n\n')}\n\n`;
};

export const updateVisualCvTex = (originalTex: string, papers: PaperItem[]): string => {
    const newPubSection = generatePublicationsLatex(papers);

    const pattern = /(\\section\{Publication\}[\s\S]*?)(?=(?:% =+\s*\n\s*% CERTIFICATIONS|\\section\{Certifications\}))/;
    const match = pattern.exec(originalTex);

    if (match) {
        return originalTex.slice(0, match.index) + newPubSection + originalTex.slice(match.index + match[0].length);
    }

    // Fallback: search for \section{Publication} up to next \section{
    const fallbackPattern = /(\\section\{Publication\}[\s\S]*?)(?=\\section\{)/;
    const fallbackMatch = fallbackPattern.exec(originalTex);
    if (fallbackMatch) {
        return originalTex.slice(0, fallbackMatch.index) + newPubSection + originalTex.slice(fallbackMatch.index + fallbackMatch[0].length);
    }

    return originalTex;
};

export const downloadUpdatedVisualCvTex = async (
    papers: PaperItem[],
    texPath: string = '/resume/Shah_Abdul_Mazid_Visual_CV_Version_2.tex',
    filename: string = 'Shah_Abdul_Mazid_Visual_CV_Version_2.tex'
): Promise<void> => {
    try {
        const res = await fetch(texPath);
        if (!res.ok) throw new Error(`HTTP error ${res.status}`);
        const originalTex = await res.text();
        const updatedTex = updateVisualCvTex(originalTex, papers);

        const blob = new Blob([updatedTex], { type: 'text/plain;charset=utf-8' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = filename;
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
        URL.revokeObjectURL(url);
    } catch (err) {
        console.error('Failed to generate updated .tex file, downloading original:', err);
        const a = document.createElement('a');
        a.href = texPath;
        a.download = filename;
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
    }
};

File: Portfolio_Backend/app/utils/latex_sync.py
import os
import re
import logging

logger = logging.getLogger("latex-sync")

def escape_latex(text: str) -> str:
    """Escapes special LaTeX characters for regular text."""
    if not text:
        return ""
    text = re.sub(r'(?<!\\)&', r'\&', text)
    text = re.sub(r'(?<!\\)%', r'\%', text)
    text = re.sub(r'(?<!\\)#', r'\#', text)
    text = re.sub(r'(?<!\\)\$', r'\$', text)
    return text

def format_authors_latex(authors_str: str) -> str:
    """Formats author string bolding Shah Abdul Mazid."""
    if not authors_str:
        return r"\textbf{Shah Abdul Mazid}"
    
    clean = escape_latex(authors_str)
    
    patterns = [
        r'(?<!\\textbf\{)Shah Abdul Mazid(?!\})',
        r'(?<!\\textbf\{)Mazid, S\.\s*A\.(?!\})',
        r'(?<!\\textbf\{)Mazid, S\.A\.(?!\})',
    ]
    for pat in patterns:
        clean = re.sub(pat, r'\\textbf{Shah Abdul Mazid}', clean)
        
    return clean

def format_paper_latex(paper: dict) -> str:
    """Converts a single paper dict to LaTeX CV entry."""
    title = escape_latex(paper.get("title", "").strip())
    authors = format_authors_latex(paper.get("authors", "").strip())
    venue = escape_latex(paper.get("venue", "").strip())
    publisher = escape_latex(paper.get("publisher", "").strip())
    year = str(paper.get("year", "")).strip()
    doi = str(paper.get("doi", "")).strip()
    link = str(paper.get("link", "")).strip()

    lines = [f"\\textbf{{{title}}}\\\\"]
    
    if authors:
        lines.append(f"\\textbf{{Authors:}} {authors}\\\\[-0.5mm]")
        
    venue_parts = []
    if venue:
        venue_parts.append(f"\\textit{{{venue}}}")
    if publisher:
        venue_parts.append(publisher)
    if year and year not in venue and year not in publisher:
        venue_parts.append(year)
        
    if venue_parts:
        lines.append(f"{', '.join(venue_parts)}.\\\\[-1mm]")
        
    if doi:
        clean_doi = doi.replace("https://doi.org/", "").strip()
        escaped_doi = clean_doi.replace("_", r"\_")
        lines.append(f"DOI: \\href{{https://doi.org/{clean_doi}}}\n{{{escaped_doi}}}")
    elif link:
        lines.append(f"Link: \\href{{{link}}}{{View Publication}}")
        
    return "\n".join(lines)

def generate_publications_section(papers: list) -> str:
    r"""Generates the full \section{Publication} LaTeX code for relevant papers."""
    if not papers:
        return "\\section{Publication}\n\n"
        
    mazid_papers = [
        p for p in papers 
        if not p.get("authors") 
        or "mazid" in p.get("authors", "").lower() 
        or "shah" in p.get("authors", "").lower()
    ]
    selected_papers = mazid_papers if mazid_papers else papers

    try:
        selected_papers = sorted(
            selected_papers, 
            key=lambda x: str(x.get("year", "0")), 
            reverse=True
        )
    except Exception:
        pass

    entries = [format_paper_latex(p) for p in selected_papers]
    
    latex = "\\section{Publication}\n\n" + "\n\n\\vspace{1mm}\n\n".join(entries) + "\n\n"
    return latex

def find_tex_file_path() -> str:
    """Attempts to resolve the absolute path to Shah_Abdul_Mazid_Visual_CV_Version_2.tex."""
    candidates = [
        r"C:\Users\LENOVO\Desktop\Portfolio_Final\Portfolio_Frontend\public\resume\Shah_Abdul_Mazid_Visual_CV_Version_2.tex",
        os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../Portfolio_Frontend/public/resume/Shah_Abdul_Mazid_Visual_CV_Version_2.tex")),
        os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../Portfolio_Frontend/public/resume/Shah_Abdul_Mazid_Visual_CV_Version_2.tex")),
    ]
    for path in candidates:
        if os.path.isfile(path):
            return path
    return candidates[0]

def update_visual_cv_latex(papers: list, target_path: str = None) -> bool:
    """Updates the publication section in Shah_Abdul_Mazid_Visual_CV_Version_2.tex."""
    tex_path = target_path or find_tex_file_path()
    if not os.path.isfile(tex_path):
        logger.warning(f"LaTeX file not found at {tex_path}")
        return False

    try:
        with open(tex_path, "r", encoding="utf-8") as f:
            content = f.read()

        new_pub_section = generate_publications_section(papers)

        pattern = r"(\\section\{Publication\}[\s\S]*?)(?=(?:% =+\s*\n\s*% CERTIFICATIONS|\\section\{Certifications\}))"
        match = re.search(pattern, content)
        
        if match:
            updated_content = content[:match.start()] + new_pub_section + content[match.end():]
        else:
            pattern_fallback = r"(\\section\{Publication\}[\s\S]*?)(?=\\section\{)"
            matches = list(re.finditer(pattern_fallback, content))
            if matches:
                m = matches[0]
                updated_content = content[:m.start()] + new_pub_section + content[m.end():]
            else:
                logger.error("Could not locate \\section{Publication} in LaTeX file.")
                return False

        with open(tex_path, "w", encoding="utf-8") as f:
            f.write(updated_content)

        logger.info(f"Successfully updated LaTeX publications at {tex_path}")
        return True

    except Exception as e:
        logger.error(f"Error updating LaTeX file: {e}")
        return False
